src/txmempool.cpp

```diff
@@ -684,31 +652,80 @@ void CTxMemPool::removeWithAnchor(const uint256 &invalidRoot)
 
 void CTxMemPool::removeConflicts(const CTransaction &tx, std::list<CTransaction>& removedTxs, std::list<CScCertificate>& removedCerts)
 {
-    // Remove transactions which depend on inputs of tx, recursively
-    // not used
-    // list<CTransaction> result;
     LOCK(cs);
-    BOOST_FOREACH(const CTxIn &txin, tx.GetVin()) {
+
+    for(const CTxIn &txin: tx.GetVin())
+    {
         std::map<COutPoint, CInPoint>::iterator it = mapNextTx.find(txin.prevout);
-        if (it != mapNextTx.end()) {
-            const CTransactionBase &txConflict = *it->second.ptx;
+        if (it == mapNextTx.end())
+            continue;
+
+        const CTransactionBase &txConflict = *it->second.ptx;
+        if (txConflict != tx)
+            remove(txConflict, removedTxs, removedCerts, true);
+    }
+
+    for(const JSDescription &joinsplit: tx.GetVjoinsplit())
+    {
+        for(const uint256 &nf: joinsplit.nullifiers)
+        {
+            std::map<uint256, const CTransaction*>::iterator it = mapNullifiers.find(nf);
+            if (it == mapNullifiers.end())
+                continue;
+
+            const CTransactionBase &txConflict = *it->second;
             if (txConflict != tx)
-            {
                 remove(txConflict, removedTxs, removedCerts, true);
-            }
+
         }
     }
+}
 
-    BOOST_FOREACH(const JSDescription &joinsplit, tx.GetVjoinsplit()) {
-        BOOST_FOREACH(const uint256 &nf, joinsplit.nullifiers) {
-            std::map<uint256, const CTransaction*>::iterator it = mapNullifiers.find(nf);
-            if (it != mapNullifiers.end()) {
-                const CTransactionBase &txConflict = *it->second;
-                if (txConflict != tx)
-                {
-                    remove(txConflict, removedTxs, removedCerts, true);
-                }
-            }
+void CTxMemPool::removeStaleTransactions(const CCoinsViewCache * const pCoinsView, unsigned int nMemPoolHeight,
+                                         std::list<CTransaction>& outdatedTxs, std::list<CScCertificate>& outdatedCerts)
+{
+    LOCK(cs);
+    std::set<uint256> txesToRemove;
+
+    for (std::map<uint256, CTxMemPoolEntry>::const_iterator it = mapTx.begin(); it != mapTx.end(); it++)
+    {
+        const CTransaction& tx = it->second.GetTx();
+
+        if (!checkTxImmatureExpenditures(tx, pCoinsView, nMemPoolHeight))
+        {
+            txesToRemove.insert(tx.GetHash());
+            continue;
+        }
+    }
+
+    // mbtr will be removed if sidechain is beyond safeguard
+    for (auto it = mapSidechains.begin(); it != mapSidechains.end(); it++)
```

Comment:
this method is called both by connect and disconnect tip. We should remove mbtrs every time the current certDataHash is changed by a tip change (see comment in txmempool.h). Moreover mcbtrs must be removed in case of ceased sc.

Reply:
Introduce the check for removing fts in case of ceased scs.